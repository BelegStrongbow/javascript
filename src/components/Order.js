import React from "react";
import { defineMessages, injectIntl, intlShape, FormattedNumber, FormattedDate } from "react-intl";
import { RowResponsive, ColumnPrimary, ColumnFixedWidth, ColumnMinWidth, makeFullWidth } from "./Tables";
import { IconButtonLink, disable, IconButton, makeButtonFullWidth } from "./Button";
import MediaQuery from "react-responsive";
import downloadIcon from "../icons/download.svg";
import formatAmount from "../../../shared/currency";
import LineItems from "./LineItems";
import defaults from "../config/defaults.json";

const messages = defineMessages( {
	date: {
		id: "orders.overview.date",
		defaultMessage: "Date",
	},
	orderNumber: {
		id: "orders.overview.orderNumber",
		defaultMessage: "Order",
	},
	items: {
		id: "orders.overview.items",
		defaultMessage: "Items",
	},
	total: {
		id: "orders.overview.total",
		defaultMessage: "Total",
	},
	status: {
		id: "orders.overview.status",
		defaultMessage: "Status",
	},
	invoice: {
		id: "orders.overview.invoice",
		defaultMessage: "Invoice",
	},
	invoiceLabel: {
		id: "orders.overview.invoice.label",
		defaultMessage: "Download invoice",
	},
} );

let ColumnMinWidthResponsive = makeFullWidth( ColumnMinWidth );
let ColumnPrimaryResponsive = makeFullWidth( ColumnPrimary );
let ColumnFixedWidthResponsive = makeFullWidth( ColumnFixedWidth );

let invoiceStatuses = [ "completed", "refunded" ];

/**
 * A page order list using table abstraction.
 *
 * @param {Object} props Properties to be passed to the table.
 * @returns {ReactElement} A row of order stuff.
 */
function Order( props ) {
	let InvoiceButton = IconButtonLink;
	if ( ! invoiceStatuses.includes( props.status ) ) {
		InvoiceButton = disable( IconButton );
	}

	let ResponsiveInvoiceButton = makeButtonFullWidth( InvoiceButton );

	let invoiceMessage = props.intl.formatMessage( messages.invoice );
	let invoiceLabel = props.intl.formatMessage( messages.invoiceLabel );

	return (
		<RowResponsive background={ props.background }>
			<ColumnMinWidthResponsive ellipsis={ true } headerLabel={ props.intl.formatMessage( messages.date ) }>
				<FormattedDate value={ props.date } day="numeric" month="long" year="numeric"/>
			</ColumnMinWidthResponsive>
			<ColumnMinWidth ellipsis={ true } hideOnMobile={ true } hideOnTablet={ true }
						headerLabel={ props.intl.formatMessage( messages.orderNumber ) }>
				{ props.orderNumber }
			</ColumnMinWidth>
			<ColumnPrimaryResponsive ellipsis={ true } headerLabel={ props.intl.formatMessage( messages.items ) }>
				<LineItems items={ props.items }/>
			</ColumnPrimaryResponsive>
			<ColumnMinWidthResponsive ellipsis={ true } headerLabel={ props.intl.formatMessage( messages.total ) }>
				<FormattedNumber value={ formatAmount( props.total ) } style="currency" currency={ props.currency }/>
			</ColumnMinWidthResponsive>
			<ColumnMinWidthResponsive ellipsis={ true } headerLabel={ props.intl.formatMessage( messages.status ) }>
				{ props.status }
			</ColumnMinWidthResponsive>
			<ColumnFixedWidthResponsive>
				<ResponsiveInvoiceButton
					aria-label={ invoiceLabel }
					iconSource={ downloadIcon }
					to={ props.invoiceLink }>
					<MediaQuery query={ `(min-width: ${ defaults.css.breakpoint.tablet + 1 }px)` } component="span">
						{ invoiceMessage }
					</MediaQuery>
					<MediaQuery query={ `(min-width: ${ defaults.css.breakpoint.mobile + 1 }px) and (max-width: ${ defaults.css.breakpoint.tablet }px)` }>
						<span className="screen-reader-text">{ invoiceMessage }</span>
					</MediaQuery>
					<MediaQuery query={ `(max-width: ${ defaults.css.breakpoint.mobile }px)` } component="span">
						{ invoiceMessage }
					</MediaQuery>
				</ResponsiveInvoiceButton>
			</ColumnFixedWidthResponsive>
		</RowResponsive>
	);
}

Order.propTypes = {
	date: React.PropTypes.instanceOf( Date ).isRequired,
	id: React.PropTypes.string.isRequired,
	orderNumber: React.PropTypes.string.isRequired,
	total: React.PropTypes.number.isRequired,
	currency: React.PropTypes.string.isRequired,
	invoiceLink: React.PropTypes.string,
	intl: intlShape.isRequired,
	items: React.PropTypes.array,
	status: React.PropTypes.string,
	background: React.PropTypes.string,
};

Order.defaultProps = {
	items: [],
	status: "N/A",
	background: "",
};

export default injectIntl( Order );
