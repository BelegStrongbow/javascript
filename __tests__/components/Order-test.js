import React from "react";
import { createComponentWithIntl } from "../../utils";
import { MemoryRouter } from "react-router-dom";
import Order from "../../src/components/Order";

test( "The Order component matches the snapshot", () => {
	let order =
		{
			date: new Date( "01/01/2012" ),
			orderNumber: "MOOIE 456 TEST",
			items: "TEST ITEM",
			total: 10000,
			currency: "USD",
			status: "Failed",
		};

	const component = createComponentWithIntl(
		<MemoryRouter>
			<Order
				id="a"
				invoiceLink="http://somelink"
				{ ...order }
			/>
		</MemoryRouter>
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );
