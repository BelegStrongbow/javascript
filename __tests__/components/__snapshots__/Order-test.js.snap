// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A standard order 1`] = `
<li
  className="bRBeej"
>
  <span
    className="gRzsfa"
  >
    <span>
      January 1, 2012
    </span>
  </span>
  <span
    className="fKqvxj"
  >
    MOOIE 456 TEST
  </span>
  <span
    className="iSxaIr"
  >
    <ul
      className="tvpac"
    >
      <li
        className="a"
      >
        1
        x 
        Name of the product
      </li>
      <li
        className="a"
      >
        2
        x 
        Another product
      </li>
    </ul>
  </span>
  <span
    className="FOjhs"
  >
    <span>
      $100.00
    </span>
  </span>
  <span
    className="gWfuXy"
  >
    <span>
      Completed
    </span>
  </span>
  <span
    className="gJSEqq"
  >
    <button
      className="gqtpyi liRnCl"
      disabled="disabled"
      to="http://somelink"
    >
      <span
        className="screen-reader-text"
      >
        Invoice
      </span>
    </button>
  </span>
</li>
`;

exports[`An Order with a disabled invoice button 1`] = `
<li
  className="bRBeej"
>
  <span
    className="gRzsfa"
  >
    <span>
      January 1, 2012
    </span>
  </span>
  <span
    className="fKqvxj"
  >
    MOOIE 456 TEST
  </span>
  <span
    className="iSxaIr"
  >
    <ul
      className="tvpac"
    >
      <li
        className="a"
      >
        1
        x 
        Name of the product
      </li>
      <li
        className="a"
      >
        2
        x 
        Another product
      </li>
    </ul>
  </span>
  <span
    className="FOjhs"
  >
    <span>
      $100.00
    </span>
  </span>
  <span
    className="gWfuXy"
  >
    <span>
      Pending
    </span>
  </span>
  <span
    className="gJSEqq"
  >
    <button
      className="gqtpyi liRnCl"
      disabled="disabled"
      to="http://somelink"
    >
      <span
        className="screen-reader-text"
      >
        Invoice
      </span>
    </button>
  </span>
</li>
`;
